### PR TITLE
Added build-mono-docker.ps1

### PR DESCRIPTION
**Pull request by jnm2**
Via https://github.com/nunit/docs/issues/163#issuecomment-297122590

### CODE REVIEW

**Comment by ChrisMaddock** [comment]
Would you maybe mind adding a note to BUILDING.md on what this files there for? It's fairly self-explanatory in isolation - but we have a lot of build scripts!

File: build-mono-docker.ps1 [hunk 1]
```diff
@@ -0,0 +1,7 @@
+trap [Management.Automation.CommandNotFoundException] {
+  Write-Error 'Docker cannot be found. Make sure it is installed and added to the path.'
+  Start-Process -FilePath 'https://docs.docker.com/docker-for-windows/install/'
+  continue;
+}
+
+docker run --rm -it -v ${pwd}:/nunit -w=/nunit mono bash build.sh $args
```
Comment (rprouse) [inline build-mono-docker.ps1]: Not sure about this, so putting it out for discussion. Should we switch `mono` to `mono:latest` so it is explicit?

**Comment by rprouse** [comment]
It was mentioned elsewhere, but I had to delete `build.sh` from my source directory and check it out again for the line endings to switch to `LF`. Other than that, this worked well, thanks. Every time I play with Docker, I like it more 😄

File: build-mono-docker.ps1 (hunk 1 above)
Comment (jnm2) [inline build-mono-docker.ps1]: Yes, sounds best.

**Comment by jnm2** [comment]
How is that? /cc @CharliePoole

**Comment by CharliePoole** [comment]
I'm OK with merging this script as a start... I see some issues...

1. For anyone who has not used Docker before (like me) it's a fairly steep learning curve just running the script dealing with each error message that comes up. Things I had to do to run it...
  * Install Docker for Windows
  * Read the docs and figure out how to set it up and use it
  * Enable VtX in my bios
I don't think your instructions should explain all this stuff, but it should point people to the info - or just tell them not to try to use it without first gaining some mastery of Docker.

2. I'd rather see the instructions in a separate file from Building.md. As we all know, I would prefer a text file, but you guys are all in love with markdown. 😄 

3. Personally, I would prefer to simply run bash in a linux+mono environment and then issue issue the appropriate build commands myself rather than going through the build-mono-docker command each time. I can type Exit when I'm done.

4. The script name seems too long to me.

5. We should emulate docker itself and have a .cmd script as well as a .ps1 script.

**Comment by jnm2** [comment]
1. Docker was extremely plug and play for me. Besides double-clicking the installer there was only one step which I left out, that you have to go to Settings > Shared Drives and tick the box next to whichever drive the repo is on. I left it out of BUILDING.md because 1) I figured it would just be noise to most people that use Docker, unrelated to NUnit and 2) the error message if you leave this step out is so simple and descriptive you can't go wrong.
VT-x makes sense. I already had it on so I assume it gave a similarly helpful error message? Anyone who has run Android emulators or VirtualBox or Hyper-V has already set this bit.

    So if I mention that and I mention the VT-x bit, that covers all known setup contingencies so far? At least that way people will know this could involve a reboot to the BIOS.

 2. Yeah, why are we so in love with markdown? Weird. 😁 I have no problem with this, but what would I name it? (Also, I dislike adding files starting with `b` because I already have to type <kbd>b</kbd> <kbd>Tab</kbd> <kbd>Tab</kbd> <kbd>Tab</kbd> <kbd>Tab</kbd> to autocomplete `build.cmd` in the terminal. See point 4.)

 3. Not a bad idea. @rprouse What do you think of this?

 4. Suggestions for the name? `build` is because it matches the other build scripts, `mono` is so that you know why you'd want to use the file, and  `docker` is because there are other ways to build for Mono on Linux from Windows than Docker which we may add, like WSL. Right now all it takes is typing `build-m` <kbd>Tab</kbd>- does that alleviate the problem? Or see point 3- maybe we will end up not having a build script that mirrors the arguments and behavior of `build.ps1`. (@rprouse?)

 5. So just copy `build.cmd`? `@powershell ./build-mono-docker.ps1 %CAKE_ARGS% %*`

**Comment by CharliePoole** [comment]
@jnm2 I like to think about the potential audiences for a script or doc. In this case...

For hard-core windows-only folks, separate the doc, don't care about the command.  

For newbies to docker and maybe Linux as well, make a simple command that does everything. I think that's what you have, if the doc gets beefed up.

For those who are already familiar with the tech, or for newbies who decide to learn more, just tell them what containers are available. This could be at the end of the doc, for example. Some folks will want to vary the mono build and possibly the distro.

1. Between us, I think we listed most of the points to add. One more: I spent a silly amount of time figuring out where the docker settings are!

2. DOCKER-BUILD ?

3. I think that would be a different script for a different audience or they would just type a docker command.

4. Not shorter but docker-mono-build is a more natural word order for me and works better with intellisense.

5. Or just use the identical docker command and don't display the site.

**Comment by rprouse** [comment]
I approved changes based on my original requests. I think we have a few of @CharliePoole's requests to deal with now, then we can merge.

**Info on setting up Docker** - I don't think we should add too much here, maybe just point to the documentation because it is fairly new for most developers. We don't include information on setting up Visual Studio, Mono or .NET Core.

**Location of build info** - I am fine with `BUILDING.md` or in a new file. I would lean towards `BUILDING.md` because we already have so many files in the root, but I am fine with `DOCKER-BUILD.md` too.

**Using Markdown** - I just like it because it displays nicely in GitHub. I often explore files like `BUILDING.md` on the web in GitHub rather than locally, but maybe that is just me 😄 

**Name of the build file** - `docker-mono-build` 👍 

**Interactive session** - I like both ideas of just dropping into an interactive bash session, or passing the CAKE arguments.  For larger changes I work in stages running individual build tasks to save time.

**Comment by CharliePoole** [comment]
The CAKE_ARGS were intended for local use. Running in a container is "sort of" local, but also a bit different. Also, the args I used on my old windows system dealt with a problem of Cake on 32-bit Windows systems. I don't know what may be needed under Linux, but it's likely to be different. Until we know more, I wouldn't pass them.

**Comment by CharliePoole** [comment]
I think this is going to develop over time so I'm OK with it as is or with any of the changes we discussed. I don't think changing our build scripts in the future is an issue.

@jnm2 I think you should consider who your primary audience is. Either satisfy those who know a lot about docker or those who know nothing and just want to run a quick test. Make what changes make sense to you and then merge.

Here's one point of view from an old-timer. 😄 When I'm working at the command line, I expect to see text. I don't expect windows or web pages to pop up. I don't like surprises. Ideally, our build command would have a help option displaying text, but I never added it. That said, I'm pretty sure I'm not your typical audience.

**Comment by CharliePoole** [comment]
It may be good to mention in the doc that we are using mono:latest under Debian 7 (Wheezy) for those who care.

**Comment by jnm2** [comment]
Oops- I did not get around to responding to everyone's feedback here. I guess anyone can feel free to improve the setup in their own PR. =)

**Comment by rprouse** [comment]
No worries, it is good enough and it works.

**Pull request merged**